File: python/socketd/socketd/transport/core/entity/EntityDefault.py
import os

from io import BytesIO, TextIOWrapper, BufferedReader
from typing import Any, Optional, BinaryIO

from socketd.transport.core.Entity import Entity
from socketd.transport.core.Costants import Constants
from socketd.transport.core.EntityMetas import EntityMetas


class EntityDefault(Entity):
    def __init__(self):
        self._meta_map: Optional[dict] = None
        self._meta_string = Constants.DEF_META_STRING
        self._meta_stringChanged = False
        self._data: BytesIO | TextIOWrapper = Constants.DEF_DATA
        self._data_size = 0

    def at(self, name: str):
        self.meta_put("@", name)
        return self

    def range(self, start: int, stop: int):
        self.meta_put(EntityMetas.META_RANGE_START, start)
        self.meta_put(EntityMetas.META_RANGE_SIZE, stop)
        self._meta_stringChanged = True
        return self

    def meta_string_set(self, meta_string):
        self._meta_map = None
        self._meta_string = meta_string
        self._meta_stringChanged = False
        return self

    def meta_string(self):
        if self._meta_stringChanged:
            buf = ""
            for name, val in self.meta_map().items():
                buf += f"{name}={val}&"
            if len(buf) > 0:
                buf = buf[:-1]
            self._meta_string = buf
            self._meta_stringChanged = False
        return self._meta_string

    def meta_map_put(self, metaMap:dict[str,str]):
        if metaMap:
            self.meta_map().update(metaMap)
            self._meta_stringChanged = True
        return self

    def meta_map(self):
        if self._meta_map is None:
            self._meta_map = {}
            self._meta_stringChanged = False
            if self._meta_string:
                for kv_str in self._meta_string.split("&"):
                    kv = kv_str.split("=")
                    if len(kv) > 1:
                        self._meta_map[kv[0]] = kv[1]
                    else:
                        self._meta_map[kv[0]] = ""
        return self._meta_map

    def meta_put(self, name:str, val:str):
        if val:
            self.meta_map()[name] = val
        else:
            self.meta_map().pop(name)

        self._meta_stringChanged = True
        return self

    def meta_del(self, name:str):
        self.meta_map().pop(name)
        self._meta_stringChanged = True
        return self

    def meta(self, name:str) -> Any:
        return self.meta_map().get(name)

    def meta_or_default(self, name:str, default_val:str):
        if data := self.meta_map().get(name):
            return data
        return default_val

    def meta_as_int(self, name:str) ->int:
        return int(self.meta_or_default(name,"0"))

    def put_meta(self, name:str, val:str):
        self.meta_put(name,val)

    def del_meta(self, name:str):
        self.meta_del(name)

    def data_set(self, data: bytes | bytearray | memoryview | BytesIO | BinaryIO):
        _type = type(data)
        if _type == BytesIO:
            self._data = data
            self._data_size = len(data.getvalue())
        elif _type == BufferedReader:
            self._data = data
            self._data_size = data.seek(0, os.SEEK_END)
            data.seek(0)
        else:
            self._data = BytesIO(data)
            self._data_size = len(data)
        return self

    def data(self):
        return self._data

    def data_as_string(self):
        return self._data.getvalue().decode('utf-8')  # _assuming _data is of type bytes

    def data_as_bytes(self) -> bytes:
        return self._data.getvalue()

    def data_size(self):
        return self._data_size

    def release(self):
        ...

    def __str__(self):
        return f"Entity(meta='{self.meta_string()}', data=byte[{self._data_size}])"

File: javascript/socketd/src/socketd/transport/core/entity/EntityDefault.ts
import {BlobBuffer, Buffer, ByteBuffer} from "../Buffer";
import {ArrayBufferCodecReader, CodecReader} from "../Codec";
import {Constants} from "../Constants";
import {EntityMetas} from "../EntityMetas";
import {SocketDException} from "../../../exception/SocketDException";
import {StrUtils} from "../../../utils/StrUtils";
import {Entity} from "../Entity";

/**
 * 实体默认实现
 *
 * @author noear
 * @since 2.0
 */
export class EntityDefault implements Entity {
    private _metaMap: Map<string, string> | null;
    private _data: Buffer;
    private _dataAsReader: CodecReader | null;

    constructor() {
        this._metaMap = null;
        this._data = Constants.DEF_DATA;
        this._dataAsReader = null;
    }

    /**
     * At player name
     */
    at(name: string): EntityDefault {
        this.metaPut("@", name);
        return this;
    }

    /**
     * Range
     * */
    range(start: number, size: number): EntityDefault {
        this.metaPut(EntityMetas.META_RANGE_START, start.toString());
        this.metaPut(EntityMetas.META_RANGE_SIZE, size.toString());
        return this;
    }

    /**
     * 设置元信息字符串
     * */
    metaStringSet(metaString: string): EntityDefault {
        this._metaMap = new Map<string, string>();

        //此处要优化
        if (metaString) {
            for (const kvStr of metaString.split("&")) {
                const idx = kvStr.indexOf('=');
                if (idx > 0) {
                    this._metaMap.set(kvStr.substring(0, idx), kvStr.substring(idx + 1));
                }
            }
        }

        return this;
    }

    /**
     * 放置元信息字典
     *
     * @param map 元信息字典
     */
    metaMapPut(map: any): EntityDefault {
        if (map) {
            if (map instanceof Map) {
                map.forEach((val, key, p) => {
                    this.metaMap().set(key, val);
                })
            } else {
                for (const name of map.prototype) {
                    this.metaMap().set(name, map[name]);
                }
            }
        }

        return this;
    }

    /**
     * 放置元信息
     *
     * @param name 名字
     * @param val  值
     */
    metaPut(name: string, val: string | null): EntityDefault {
        if (val == null) {
            this.metaMap().delete(name);
        } else {
            this.metaMap().set(name, val);
        }

        return this;
    }

    /**
     * 删除元信息
     *
     * @param name 名字
     */
    metaDel(name:string){
        this.metaMap().delete(name);
    }

    /**
     * 获取元信息字符串（queryString style）
     */
    metaString(): string {
        let str = "";
        this.metaMap().forEach((val, key, p) => {
            str += `${key}=${val}&`;
        });
        if (str.length > 0) {
            return str.substring(0, str.length - 1);
        } else {
            return str;
        }
    }

    /**
     * 获取元信息字典
     */
    metaMap(): Map<string, string> {
        if (this._metaMap == null) {
            this._metaMap = new Map<string, string>();
        }

        return this._metaMap;
    }

    /**
     * 获取元信息
     *
     * @param name 名字
     */
    meta(name: string): string | null {
        let tmp = this.metaMap().get(name);
        return tmp ? tmp : null;
    }

    /**
     * 获取元信息或默认值
     *
     * @param name 名字
     * @param def  默认值
     */
    metaOrDefault(name: string, def: string): string {
        const val = this.meta(name);
        if (val) {
            return val;
        } else {
            return def;
        }
    }

    /**
     * 获取元信息并转为 int
     */
    metaAsInt(name: string): number {
        return parseInt(this.metaOrDefault(name, '0'));
    }

    /**
     * 获取元信息并转为 long(int)
     */
    metaAsLong(name: string): number {
        return this.metaAsInt(name);
    }

    /**
     * 获取元信息并转为 float
     */
    metaAsFloat(name: string): number {
        return parseFloat(this.metaOrDefault(name, '0'));
    }

    /**
     * 获取元信息并转为 double(float)
     */
    metaAsDouble(name: string): number {
        return this.metaAsFloat(name);
    }

    /**
     * 放置元信息
     *
     * @param name 名字
     * @param val  值
     */
    putMeta(name: string, val: string| null) {
        this.metaPut(name, val);
    }

    /**
     * 删除元信息
     *
     * @param name 名字
     */
    delMeta(name: string) {
        this.metaDel(name);
    }

    /**
     * 设置数据
     *
     * @param data 数据
     */
    dataSet(data: Blob | ArrayBuffer | Buffer): EntityDefault {
        if (data instanceof BlobBuffer || data instanceof ByteBuffer) {
            this._data = data;
        } else if (data instanceof ArrayBuffer) {
            this._data = new ByteBuffer(data);
        } else if (typeof (Blob) != 'undefined' && data instanceof Blob) {
            this._data = new BlobBuffer(data);
        } else {
            console.warn("This data type is not supported, type=" + typeof (data));
        }

        return this;
    }

    /**
     * 获取数据（若多次复用，需要reset）
     */
    data(): Buffer {
        return this._data;
    }

    dataAsReader(): CodecReader {
        if (this._data.getArray() == null) {
            throw new SocketDException("Blob does not support dataAsReader");
        }

        if (!this._dataAsReader) {
            this._dataAsReader = new ArrayBufferCodecReader(this._data.getArray()!);
        }

        return this._dataAsReader;
    }

    /**
     * 获取数据并转成字符串
     */
    dataAsString(): string {
        if (this._data.getArray() == null) {
            throw new SocketDException("Blob does not support dataAsString");
        }

        return StrUtils.bufToStrDo(this._data.getArray()!, '');
    }

    /**
     * 获取数据长度
     */
    dataSize(): number {
        return this._data.size();
    }

    /**
     * 释放资源
     */
    release() {

    }

    toString(): string {
        return "Entity{" +
            "meta='" + this.metaString() + '\'' +
            ", data=byte[" + this.dataSize() + ']' + //避免内容太大，影响打印
            '}';
    }
}